### PR TITLE
Parse query comment and use as bigquery job labels.

### LINKED ISSUE #2483: Support BigQuery Job Tags and Labels
### Describe the feature

I would like to be able to control [tagging and labeling of BigQuery Jobs](https://cloud.google.com/bigquery/docs/adding-labels#job-label) as I run dbt on BigQuery.

A similar (but not the same) issue is #1947, for labeling BigQuery Tables and Datasets. This issue focuses on BigQuery Jobs (such as Insert Jobs or Query Jobs).

### Describe alternatives you've considered

It's not possible to label or tag jobs after they have started. From [the docs](https://cloud.google.com/bigquery/docs/adding-labels#job-label)

> You cannot add labels to or update labels on pending, running, or completed jobs.

### Additional context

The main reason why one would tag and label their BigQuery Job is to analyze BigQuery spend. For example, if one were able to link a BigQuery Job to a certain Airflow operator run (or similar -- in my case a python script run by a cron! :-D) then a real dollar value can be put on running that operator over time.

I think it's important to give the developer control on what tags and labels can be added, so it supports their data processing setup. And so I think tags and labels should be able to be set at launch-time. (In my case, I run a python script that calls `dbt run` -- I would want my python script to be able to set the BigQuery Job tags and labels, while the Jobs are ultimately launched by `dbt run`.)

### Who will this benefit?

Folks who are responsible for their BigQuery spend should benefit by using relevant Job tags and labels.

Thanks!

### PR DESCRIPTION
**Pull request by jmcarp**
resolves #2483

### Description

Try to parse query comment and pass values to bigquery job labels. Just a draft to see if I'm heading in the right direction. @jtcohen6 

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: core/dbt/contracts/connection.py
Comment (jtcohen6) [inline core/dbt/contracts/connection.py:218]: The config is `query-comment`, rather than `query_comment` ([docs](https://docs.getdbt.com/reference/project-configs/query-comment)). Not exactly sure why we did this, but I guess kebab casing is common in `dbt_project.yml` / project-level configs. I think this should be `job-label` (instead of `job_label`) for consistency

File: plugins/bigquery/dbt/adapters/bigquery/connections.py
Comment (jtcohen6) [inline plugins/bigquery/dbt/adapters/bigquery/connections.py:325]: If the comment isn't valid json, should we still include it as a single job label (called `query_comment`), truncated to 128 characters if need be?

**Review by jtcohen6 — COMMENTED** [review]
This is really cool @jmcarp!

Screen Shot 2021-03-04 at 12 33 39 PM

Thinking about how to test this:
- A unit test could inspect `job_params[labels]` and confirms that they contain the keys from the default query comment (`app`, `profile_name`, `target_name`, `dbt_version`, `node_id`) and the `dbt_invocation_id`. Maybe the query comment-to-job label conversion should be a standalone function?
- An integration test could run a model and check that the labels appropriately registered, by querying `INFORMATION_SCHEMA.JOBS_BY_USER`

File: core/dbt/contracts/connection.py
Comment (jmcarp) [reply core/dbt/contracts/connection.py:218]: OK, I switched the base class to `HyphenatedDbtClassMixin`.

Comment (jmcarp) [reply core/dbt/contracts/connection.py:218]: On second thought, I may not understand what this base class does, since that change broke tests that I thought were unrelated. Reverted for now, but let me know if you have thoughts about the right way to do this.

File: core/dbt/dataclass_schema.py
Comment (jmcarp) [inline core/dbt/dataclass_schema.py:78]: Does this make sense @gshank?

File: core/dbt/contracts/connection.py
Comment (jmcarp) [reply core/dbt/contracts/connection.py:218]: I took another look at this, and I think I found a subtle bug in the casing logic. Should be fixed now.

File: plugins/bigquery/dbt/adapters/bigquery/connections.py
Comment (jtcohen6) [inline plugins/bigquery/dbt/adapters/bigquery/connections.py:602]: Anecdotally, this has the effect of appending an extra underscore at the end of my non-dict comment, e.g. `comment: whatever` in `dbt_project.yml` becomes `query_comment: whatever_` in BigQuery, and `comment: something else` becomes `query_comment: something_else_`. Is that intended?

**Review by jtcohen6 — COMMENTED** [review]
This is looking good from my vantage! One small comment from my testing. I think it'd be good to add a few unit tests of the various inputs/outputs of query comment sanitizing and structuring. Beyond that, I think this is ready for a changelog entry (v0.20.0 Feature).

I also tagged a member of the Core team for python code review.

File: plugins/bigquery/dbt/adapters/bigquery/connections.py
Comment (jmcarp) [reply plugins/bigquery/dbt/adapters/bigquery/connections.py]: I'm not sure why this is happening--there may be something upstream of this code adding trailing whitespace, because I don't get that behavior testing the function directly. I tried adding a `strip()`. Does that help?

Comment (jtcohen6) [reply plugins/bigquery/dbt/adapters/bigquery/connections.py]: Looks like the `strip()` fixed it!

Screen Shot 2021-03-09 at 10 39 33 AM

(The numerous `_` in the top example are all special characters.)

**Review by jtcohen6 — APPROVED** [review]
This looks awesome @jmcarp! Thanks for adding those unit tests.

Functional review is good by me: I've confirmed this works as expected when `job-label: True` (with default, custom JSON, and custom string comments), and just as crucially that it works as expected when `job-label: False` or not supplied (the only label is `dbt_invocation_id` = status quo).

I'll leave final code/style review to @nathaniel-may. After that, this is good to go.

**Comment by jmcarp** [comment]
No rush @jtcohen6, but it would be awesome to get this into the next release.

File: core/dbt/dataclass_schema.py
Comment (nathaniel-may) [reply core/dbt/dataclass_schema.py:78]: Looking at all the existing calls to this class method, it seems to be true.

File: test/unit/test_bigquery_adapter.py
Comment (nathaniel-may) [inline test/unit/test_bigquery_adapter.py:946]: We don't seem to use this style of test very often, but I actually really like it so I'm happy to introduce it to this file too.

File: test/unit/test_bigquery_adapter.py
Comment (nathaniel-may) [inline test/unit/test_bigquery_adapter.py:26]: I won't block the merge on this, but I'd rather put this import up just above `import unittest`.

File: plugins/bigquery/dbt/adapters/bigquery/connections.py
Comment (nathaniel-may) [inline plugins/bigquery/dbt/adapters/bigquery/connections.py:601]: Is this the behavior we want for long labels? Alternatives include raising our own error, or sending something we know is too long to big query and raising that db exception. As much as this will not often, if ever, be attempted on purpose, a typo could cause this which might be a slightly annoying user experience.

**Review by nathaniel-may — APPROVED** [review]
This was waiting on me @jmcarp. Thank you for your patience. Once we address my relatively minor questions here, this looks good to merge.

I do want an integration test so that we can detect regressions on this behavior from both dbt and bigquery, but I'm ok ticketing that in a separate PR if you are too @jtcohen6.

File: plugins/bigquery/dbt/adapters/bigquery/connections.py
Comment (jmcarp) [reply plugins/bigquery/dbt/adapters/bigquery/connections.py:601]: I think passing through the input and letting bigquery crash makes sense.

File: core/dbt/dataclass_schema.py
Comment (jmcarp) [reply core/dbt/dataclass_schema.py:78]: Honestly, I'm not sure exactly how this code gets reached, but I dropped into a debugger at this point on the unit test suite and found non-dictionary values--specifically, there's a test that sets `query-comment` to `""`. You should be able to reproduce the issue by dropping the `isinstance` check that I added and running something like `pytest -s test/unit/test_query_headers.py -k test_disable_query_comment`.

Comment (jmcarp) [reply core/dbt/dataclass_schema.py:78]: You can also run that test with the following added to __pre_deserialize__:

```
        if cls._hyphenated and not isinstance(data, dict):
            import ipdb; ipdb.set_trace()
```

Looking at the stack, I'm inside a function called `__unpack_union_Project_query_comment__9f5a8d0e89384cc286d2b99acef5623d` that I'm guessing is dynamically generated and eval-ed by mashumaro 😱.

**Comment by jtcohen6** [comment]
Thanks for the review @nathaniel-may, and for the quick cleanup @jmcarp! I'll open a separate issue for the integration test.

**Pull request merged**
